File: B13/Subhendu/Array/pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        ## https://leetcode.com/problems/pascals-triangle/
        ans = []
        for i in range(numRows):
            curr = [None] * (i + 1)
            curr[0], curr[-1] = 1, 1

            for j in range(1, i):
                curr[j] = ans[i - 1][j - 1] + ans[i - 1][j]
            ans.append(curr)
        return ans

File: B13/Array/Hemanth/Pascals_triangle.java
//https://leetcode.com/problems/pascals-triangle/


class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> cur, pre = null;
        for(int i=0;i<numRows;i++){
            cur = new ArrayList<>();
            for(int j=0;j<=i;j++){
            if(j==0 || j==i){
                cur.add(1);
            }else{
                cur.add(pre.get(j-1)+pre.get(j));
            }
            }
            pre = cur;
        res.add(cur);
        }
        
        return res;
    }
}

//Time Complexity : O(n^2)
//Space Complexity : O(n^2)
